[csharp tools] Option to append Async to server side method names #39010

ganesh-tn opened the pull request:
Issue #39010 - Option to append Async to server side method names

ganesh-tn committed:
Option to append Async to server side method names #39010

ganesh-tn commented:
The default setting is enabled (true). To disable this behavior, please specify the option `--grpc_opt=appendasyncsuffix=false`.

apolcyn reviewed (commented):
Thanks @ganesh-tn 

A couple first comments:

Especially since this is a new feature, can we default this off?

Can we add test coverage, either in the form of an example (e..g under the examples directory), or as a unit test? (e.g. under `Grpc.Tools.Tests` directory)

ganesh-tn committed:
Default settings and examples

- The default option is set to false.
- An example file has been attached for reference.

apolcyn commented on examples/csharp/AsyncSuffix/Greet/generate_proto.sh:
```diff
@@ -0,0 +1,9 @@
+protoc --csharp_out=./Server/ --grpc_out=./Server/ --plugin=protoc-gen-grpc=grpc_csharp_plugin greet.proto
+
+# Generate with async suffix
+protoc --csharp_out=./Server/ --grpc_out=./Server/ --plugin=protoc-gen-grpc=grpc_csharp_plugin --grpc_opt=append_async_suffix=true greet.proto
```
since this directory is focused on the async suffix option, let's just generate the protos with the async suffix option

apolcyn commented on examples/csharp/AsyncSuffix/Server/Program.cs:
```diff
@@ -0,0 +1,58 @@
+    class GreeterService : Greeter.GreeterBase
+    {
+        // Standard unary method
+        public override Task<HelloReply> SayHello(HelloRequest request, ServerCallContext context)
```
I might be missing something, but is this file  generated without the async suffix option?

Let's check in the file that's generated with the async suffix option.

ganesh-tn replied:
The example files have been updated and corrected.

apolcyn commented on src/compiler/csharp_generator.cc:
```diff
@@ -465,14 +465,22 @@ void GenerateServerClass(Printer* out, const ServiceDescriptor* service) {
     const MethodDescriptor* method = service->method(i);
+    std::string method_name(method->name());
+    if (append_async_suffix == true) {
+      if (!method->client_streaming() && !method->server_streaming() && method_name.size() >= 5 &&
```
Just curious, what's the motivation for not including `Async` suffix on non-unary RPC handlers?

Going by https://github.com/grpc/grpc/issues/39010 and https://github.com/grpc/grpc/issues/7874, my first intuition is that the `Async` suffix to be applied whenever the protoc plugin option is set.

Also, cc @JamesNK any thoughts?

ganesh-tn replied:
- In gRPC C#, only unary RPCs use the `Async` suffix to reflect their `Task<T>` based async pattern. 
- Because the streaming interfaces already imply asynchronous usage, appending `Async` to their method names was seen as redundant or unnecessary. 
- Client streaming, server streaming, and bidirectional streaming methods return streaming call objects like `AsyncClientStreamingCall` or `AsyncServerStreamingCall`, which encapsulate asynchronous.

@apolcyn should the `async` keyword be applied to all methods within this section?

JamesNK replied:
What is the goal here?

Is the goal to make the server method names match the client ones? Only the async unary call in the client has an async suffix, because it is the only method that returns `Task`.

If `Async` suffix is added to gRPC server methods then it should be added to all of them. They all return `Task` and so are (potentially) async methods.

apolcyn reviewed (commented):
looks like a good start, some comments

TN-GANESH committed:
Modified the example files

The example files have been updated and corrected.

ganesh-tn committed:
#39010 Option to append Async to server side method names

apolcyn commented on src/compiler/csharp_generator.cc:
```diff
@@ -701,9 +709,17 @@ void GenerateBindServiceMethod(Printer* out, const ServiceDescriptor* service) {
     const MethodDescriptor* method = service->method(i);
+    std::string method_name(method->name());
+    if (append_async_suffix == true) {
+      if (!method->client_streaming() && !method->server_streaming() && method_name.size() >= 5 &&
```
I might be mis-reading -- shouldn't we use `Async` suffix on all four RPC types?

ganesh-tn committed:
Option to append Async to server side method names

ganesh-tn committed:
Option to append Async to server side method names

ganesh-tn committed:
Sanity testing - fixed

ganesh-tn committed:
Added "Async Suffix" Test Case

ganesh-tn committed:
Added "Async Suffix" Test Case

ganesh-tn committed:
Fixed "Async Suffix" Test Case

asheshvidyut reviewed (approved):
LGTM

ganesh-tn committed:
Fix - Protobuf Symbol Conflict

asheshvidyut committed:
Merge branch 'master' into PR-Ganesh-TN-39010-Option-to-append-Async-to-server-side-method-names
